horovod/torch/sync_batch_norm.py

```diff
@@ -0,0 +1,187 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from horovod.torch.mpi_ops import allgather_async, allreduce_async, Sum, size, synchronize
+
+from distutils.version import LooseVersion
+
+import torch
+from torch.autograd.function import Function
+import torch.nn.functional as F
+from torch.nn.modules.batchnorm import _BatchNorm
+
+
+# Backward compat for old PyTorch
+if not hasattr(torch.jit, 'unused'):
+    torch.jit.unused = lambda x: x
+
+
+class SyncBatchNorm(_BatchNorm):
+    """
+    Applies synchronous version of N-dimensional BatchNorm.  In this version, normalization
+    parameters are synchronized across workers during forward pass.  This is very useful in
+    situations where each GPU can fit a very small number of examples.
+
+    See https://pytorch.org/docs/stable/nn.html#batchnorm2d for more details about BatchNorm.
+
+    Arguments:
+        num_features: number of channels `C` from the shape `(N, C, ...)`
+        eps: a value added to the denominator for numerical stability. Default: 1e-5
+        momentum: the value used for the running_mean and running_var
+            computation. Can be set to `None` for cumulative moving average
+            (i.e. simple average). Default: 0.1
+        affine: a boolean value that when set to `True`, this module has
+            learnable affine parameters. Default: `True`
+        track_running_stats: a boolean value that when set to `True`, this
+            module tracks the running mean and variance, and when set to `False`,
+            this module does not track such statistics and always uses batch
+            statistics in both training and eval modes. Default: `True`
+    """
+    def __init__(self, num_features, eps=1e-5, momentum=0.1, affine=True, track_running_stats=True):
+        super().__init__(num_features, eps, momentum, affine, track_running_stats)
+
+    def _check_input_dim(self, input):
+        if input.dim() < 2:
+            raise ValueError('expected at least 2D input (got {}D input)'.format(input.dim()))
+
+    def _run_bn(self, input):
+        return F.batch_norm(
+            input, self.running_mean, self.running_var, self.weight, self.bias,
+            self.training or not self.track_running_stats, self.momentum, self.eps)
+
+    @torch.jit.unused
+    def _maybe_run_sync_bn(self, input):
+        if size() == 1:
+            return self._run_bn(input)
+        return _SyncBatchNorm.apply(
+            input, self.weight, self.bias, self.running_mean, self.running_var,
+            self.eps, self.momentum)
+
+    def forward(self, input):
+        # currently only GPU input is supported
```

Comment:
Can you add a TODO stating what would be needed to get it work on CPU?

Reply:
This is constrained by PyTorch internal kernels that we use. They're only implemented on GPU.  If we want to make it work on CPU, we should consider rolling our own implementation w/o dependency on PyTorch internals.